File: learning-factory/warehouse/artifacts/projects/build-your-own-x/programming-language/the-super-tiny-compiler-javascript-54fbf20b58/job_byox_build_v1_0f6667d0257bdf71d9995b689f1b4959/attempt-001/starter/src/lexer.js
import { PebbleNotImplementedError } from "./errors.js";

export class Lexer {
  constructor(source) {
    if (typeof source !== "string") {
      throw new TypeError("Pebble source must be a string");
    }

    this.source = source;
  }

  tokenize() {
    // TODO: scan this.source, track one-based locations, skip whitespace/comments,
    // reject malformed input, and append exactly one EOF token.
    throw new PebbleNotImplementedError("tokenizer");
  }
}

File: learning-factory/warehouse/artifacts/projects/build-your-own-x/programming-language/from-source-code-to-machine-code-build-your-own-compiler-from-scratch-2877aeded6/job_byox_build_v1_7a07ec294de1cc10739ceacd586b1ffd/attempt-001/starter/minnow/lexer.py
"""Turn source characters into located tokens."""

from .model import TokenKind


KEYWORDS = {
    "let": TokenKind.LET,
    "print": TokenKind.PRINT,
    "if": TokenKind.IF,
    "else": TokenKind.ELSE,
    "while": TokenKind.WHILE,
    "true": TokenKind.TRUE,
    "false": TokenKind.FALSE,
}


def tokenize(source):
    """Return a list ending in exactly one EOF token."""
    if not isinstance(source, str):
        raise TypeError("source must be str")
    # TODO: implement scanning, maximal munch, comments, locations, and range checks.
    raise NotImplementedError("implement tokenize")
